riscv-c-api.md

```diff
@@ -174,3 +174,72 @@ For example:
 
 Do we really have none of these?  I can't figure out
 `gcc/gcc/config/riscv/riscv-builtins.c`...
+
+## Constraints for Operands of Inline Assembly Statements
+
+Mainstream compilers like GCC and Clang/LLVM support writing inline assembly
+language in C and C++ programs by `__asm__` or `asm` statement; RISC-V follows
+ the common* way of using assembly language.
+
+This section lists operands constraint that can be used with inline assembly
+statements, including both RISC-V specific and common operand constraints.
+
+
+| Constraint         |                                    | Note        |
+| ------------------ | ---------------------------------- | ----------- |
+| m                  | An address that is held in a general-purpose register with offset.      |      |
+| A                  | An address that is held in a general-purpose register.      |      |
+| r                  | General purpose register                      |      |
+| f                  | Floating-point register                       |      |
+| i                  | Immediate integer operand                     |      |
+| I                  | 12-bits signed immediate integer operand      |      |
+| K                  | 5-bits unsigned immediate integer operand     |      |
+| J                  | Zero integer immediate operand                |      |
+
+NOTE: Immediate value must be compile-time constant.
+
+### The Difference Between `m` and `A` Constraint
+
+The difference between `m` and `A` is whether the operand can have offset;
+some instructions in RISC-V do not allow offset for the address operand, like
+atomic or vector load/store instructions.
+
+Give an example to demonstrate the difference; the following program is trying
+to load value from `foo[10]` and using `m` and `A` to pass that address.
+
+```c
+int *foo;
+void bar() {
+ int x;
+ __asm__ volatile ("lw %0, %1" : "=r"(x) : "m" (foo[10]));
+ __asm__ volatile ("lw %0, %1" : "=r"(x) : "A" (foo[10]));
+}
+```
+
+Then we compile with GCC with O3 option:
+
+```shell
+$ riscv64-unknown-elf-gcc x.c -o - -O -S
+...
+bar:
+       lui    a5,%hi(foo)
+       ld     a5,%lo(foo)(a5)
+ #APP
+# 4 "x.c" 1
+       lw a4, 40(a5)
+# 0 "" 2
+ #NO_APP
+       addi   a5,a5,40
+ #APP
+# 5 "x.c" 1
+       lw a5, 0(a5)
+# 0 "" 2
+ #NO_APP
+       ret
+
+```
+
+Then we can observe that the compiler will generate the form of the base register
```

Comment:
The compiler uses an immediate offset of 40 for the `m` constraint, but for the `A` constraint uses an extra `addi` instruction instead.